release: 1.6.0 (#1969)
biomejs/biome: website/src/content/blog/biome-v1-6.mdx

---
title: Biome v1.6
description: Partial support for Astro, Svelte and Vue, extend configuration from dependencies, formatter goodies.
summary: |
  Partial support for Astro, Svelte and Vue files, both CLI and LSP.

>   Now, the `extends` field is able to pick configuration files that are exported as a dependency, very useful in monorepo and big projects.

Reviewer: ```suggestion
  Now, the `extends` field is able to pick configuration files that are exported from a dependency.
```

I've removed the "very useful in monorepo and big projects" bit, because I think it's not related to monorepos or big projects. In fact, those already had an easier time of it, because they could just use relative paths :)

  The formatter has new options.
authors:
  - ema
  - team
pubDate: 2024-03-08
coverImage:
  lightSrc: "@/assets/blog/roadmap-2024/banner-light.png"
  darkSrc: "@/assets/blog/roadmap-2024/banner-dark.png"
  alt: The brand of the project. It says "Biome, toolchain of the web"
socialImage: "@/assets/social-logo.png"
---

import { Tabs, TabItem } from '@astrojs/starlight/components';
import { Image } from "astro:assets";
import astroLint from "../../assets/blog/biome-v1-6/astro-linter.png"
import vueLint from "../../assets/blog/biome-v1-6/vue-linter.png"
import svelteLint from "../../assets/blog/biome-v1-6/svelte-linter.png"

Update Biome using the following commands:

```shell
npm install --save-dev --save-exact @biomejs/biome@latest
npx @biomejs/biome migrate
```

## Partial support for Astro, Svelte and Vue files

> In this release, we're happy to provide partial support for Astor, Svelte and Vue files. What does **partial** support mean?

Reviewer: ```suggestion
In this release, we're happy to provide partial support for Astro, Svelte and Vue files. What does **partial** support mean?
```

> While the team is working on a [unified data structure](https://github.com/biomejs/biome/discussions/1726) for HTML-ish languages, we discovered that we could provide Biome functionalities to those files with just few changes, albeit some expectations.

Reviewer: ```suggestion
While the team is working on a [unified data structure](https://github.com/biomejs/biome/discussions/1726) for HTML-ish languages, we discovered that we could provide Biome functionalities to those files with just a few changes, albeit with some limitations.
```

> This means that Biome is able to analyse the JavaScript/TypeScript portion of said files, and all features are available: formatting, linting and import sorting! Here's an example that you should expect in terms of developer experience:

Reviewer: ```suggestion
This means that Biome is able to analyze the JavaScript/TypeScript portion of said files, and all features are available: formatting, linting and import sorting! Here's an example of what you should expect in terms of developer experience:
```

<Tabs>
    <TabItem label="Astro" icon="astro">
        <div align="center">
            <Image src={astroLint} alt="Screenshot of Biome linting in action for an Astro file in VSCode" width="500"  />
        </div>
    </TabItem>
    <TabItem label="Vue" icon="seti:vue">
        <div align="center">
            <Image src={vueLint} alt="Screenshot of Biome linting in action for an Vue file in VSCode" width="500"  />
        </div>
    </TabItem>
    <TabItem label="Svelte" icon="seti:svelte">
        <div align="center">
            <Image src={svelteLint} alt="Screenshot of Biome  linting in action for an Svelte file in VSCode" width="500"  />
        </div>
    </TabItem>
</Tabs>

Make sure to read the [documentation about expectations and limitations](/internals/language-support#html-super-languages-support).

## Configuration, lighter and more powerful

### Support for `biome.jsonc`

Biome now accepts the file `biome.jsonc` as configuration! Now, you can insert all the comments you want in a configuration.

### `extends` resolves dependencies

> From this version, Biome can the `extends` property can *resolve* other configuration files that are installed and exported as a dependency.

Reviewer: ```suggestion
From this version, Biome can use the `extends` property to *resolve* other configuration files that are inside installed dependencies.
```

```json name="shared-configs/package.json"
{
  "name": "@shared-configs",
  "type": "module",
  "exports": {
    "./biome": "./biome.json"
  }
}
```

This set up allows to expose a specifier `@shared-configs/biome`, which you use inside your `biome.json` file.

```json name="project/biome.json"
{
  "extends": ["@shared-configs/biome"]
}
```

The resolution of the dependencies is powered by the library [`oxc-resolver`](https://github.com/oxc-project/oxc-resolver), one of the many libraries provided by the [OXC project](https://oxc-project.github.io/). It's battle-tested and spec compliant!

:::note
You can also export `biome.jsonc` files!

```json name="shared-configs/package.json" ins="./biome.jsonc"
{
  "name": "@shared-configs",
  "type": "module",
  "exports": {
    "./biome": "./biome.jsonc"
  }
}
```
:::

### Reduced memory footprint

We **reduced** the size our configuration **by a factor of 6.5**! This change might not have massive effects on the speed of the program, but it greatly reduced the memory used when running the CLI or the LSP.

## New formatting options

> Other than fixes, the formatter provides two new options that should improve the compatibility with Prettier parity

Reviewer: ```suggestion
Other than fixes, the formatter provides two new options that should improve the compatibility with Prettier.
```

<div class="card">
    <div class="card-column">
        <h4>With variant `auto` (default)</h4>
        The attributes are automatically formatted, and they will collapse in multiple lines only when they hit certain criteria.

        ```jsx title="file.jsx"
        <Button as="link" style="primary" href="https://example.com">
            Hit me
        </Button>
        ```
    </div>
    <div class="card-column">
        <h4>With variant `multiline`</h4>
        The attributes are always formatted on multiple lines, regardless.

        ```jsx title="file.jsx"
        <Button
            as="link"
            style="primary"
            href="https://example.com"
        >
            Hit me
        </Button>
        ```
    </div>
</div>

### Option `json.formatter.trailingComma`

Previously, Biome parser introduced an option that would allow to parse JSON and JSONC files that contained a trailing comma. This was required to ease the friction caused by other tools that
tolerate trailing commas by default (e.g. VSCode, Prettier, etc.).

> Although, our formatter wasn't as tolerant. With the option `json.formatter.trailingComma`, you can apply the same rules that `js.formatter.trailingComma` applies.

Reviewer: ```suggestion
Unfortunately, our formatter wasn't as tolerant. But with this release, we've introduced the option `json.formatter.trailingComma`. It allows you to apply the same rules as with `js.formatter.trailingComma`.
```

### Option `attributePosition`

The contributor [@octoshikari](https://github.com/octoshikari) implemented this new feature by themselves! Huge thank you for helping the Biome project.

When `formatter.attributePosition` has the value `multiline`, all attributes of HTML-ish languages (JSX/TSX as for time of writing) will be collapsed on multiple lines regardless of their numbers:

<div class="card">
    <div class="card-column">
        <h4>With variant `auto` (default)</h4>
        The attributes are automatically formatted, and they will collapse in multiple lines only when they hit certain criteria.

        ```jsx title="file.jsx"
        <Button as="link" style="primary" href="https://example.com">
            Hit me
        </Button>
        ```
    </div>
    <div class="card-column">
        <h4>With variant `multiline`</h4>
        The attributes are always formatted on multiple lines, regardless.

        ```jsx title="file.jsx"
        <Button
            as="link"
            style="primary"
            href="https://example.com"
        >
            Hit me
        </Button>
        ```
    </div>
</div>

> ### Option `json.formatter.trailingComma`

Reviewer 2: @ematipico Please forgive me for being a little paranoid about the plural trailing `s`.

I just double checked the latest nightly release. We should use `trailingCommas` everywhere for the `json` formatter, and fix the option choices in CLI: `--json-formatter-trailing-commas=<omit|allow>`, and we might want to explain why it is `trailingComma` for the `js` formatter but decide to use `trailingCommas` for the `json` formatter (to keep consistency with other options).

And maybe we should mark the old option name as deprecated and implement and recommend users to use the new option name so that we can completely remove it in the future with less friction.

Author: Ah shoot, I forgot to follow up 😆  you're right

Reviewer 2: No worries. Thank you for putting up with me :D

<Tabs>
    <TabItem label="Astro" icon="astro">
        <div align="center">
            <Image src={astroLint} alt="Screenshot of Biome linting in action for an Astro file in VSCode" width="500"  />
        </div>
    </TabItem>
    <TabItem label="Vue" icon="seti:vue">
        <div align="center">
            <Image src={vueLint} alt="Screenshot of Biome linting in action for an Vue file in VSCode" width="500"  />
        </div>
    </TabItem>
    <TabItem label="Svelte" icon="seti:svelte">
        <div align="center">
            <Image src={svelteLint} alt="Screenshot of Biome  linting in action for an Svelte file in VSCode" width="500"  />
        </div>
    </TabItem>
</Tabs>

Make sure to read the [documentation about expectations and limitations](/internals/language-support#html-super-languages-support).

## Configuration, lighter and more powerful

### Support for `biome.jsonc`

Biome now accepts the file `biome.jsonc` as configuration! Now, you can insert all the comments you want in a configuration.

### `extends` resolves dependencies

> From this version, Biome can use the `extends` property to *resolve* other configuration files that are inside installed dependencies.

Reviewer 3: ```suggestion
From this version, Biome can use the `extends` property to _resolve_ other configuration files that are inside installed dependencies.
```

Author: I don't understand the suggestion, the emitted markup is the same?

Reviewer 3: Yes. It is just more conventional to use `_it_` for italics.

{
  "name": "@shared-configs",
  "type": "module",
  "exports": {
    "./biome": "./biome.json"
  }
}
```

This set up allows to expose a specifier `@shared-configs/biome`, which you use inside your `biome.json` file.

```json name="project/biome.json"
{
  "extends": ["@shared-configs/biome"]
}
```

The resolution of the dependencies is powered by the library [`oxc-resolver`](https://github.com/oxc-project/oxc-resolver), one of the many libraries provided by the [OXC project](https://oxc-project.github.io/). It's battle-tested and spec compliant!

:::note
You can also export `biome.jsonc` files!

```json name="shared-configs/package.json" ins="./biome.jsonc"
{
  "name": "@shared-configs",
  "type": "module",
  "exports": {
    "./biome": "./biome.jsonc"
  }
}
```
:::

### Reduced memory footprint

We **reduced** the size our configuration **by a factor of 6.5**! This change might not have massive effects on the speed of the program, but it greatly reduced the memory used when running the CLI or the LSP.

## New formatting options

Other than fixes, the formatter provides two new options that should improve the compatibility with Prettier.

### Option `attributePosition`

> The contributor [@octoshikari](https://github.com/octoshikari) implemented this new feature by themselves! Huge thank you for helping the Biome project.

Reviewer 3: We could move this paragraph at the end of the section. First we explain what is the config about and then we thank the contributor. What do you think?

Author: Done! It makes sense

<div class="card">
    <div class="card-column">
        <h4>With variant `auto` (default)</h4>
        The attributes are automatically formatted, and they will collapse in multiple lines only when they hit certain criteria.

        ```jsx title="file.jsx"
        <Button as="link" style="primary" href="https://example.com">
            Hit me
        </Button>
        ```
    </div>
    <div class="card-column">
        <h4>With variant `multiline`</h4>
        The attributes are always formatted on multiple lines, regardless.

        ```jsx title="file.jsx"
        <Button
            as="link"
            style="primary"
            href="https://example.com"
        >
            Hit me
        </Button>
        ```
    </div>
</div>

### Option `json.formatter.trailingCommas`

Previously, Biome parser introduced an option that would allow to parse JSON and JSONC files that contained a trailing comma. This was required to ease the friction caused by other tools that
tolerate trailing commas by default (e.g. VSCode, Prettier, etc.).

> Unfortunately, our formatter wasn't as tolerant. But with this release, we've introduced the option `json.formatter.trailingCommas`. It allows you to apply the same rules as with `js.formatter.trailingComma`.

Reviewer 3: Are there some rationals between the use of a plural form for _json_ and singular form for _js_?

Author: Singular was a mistake, plural is the way to go. Other options use the plural. We were superficial when we introduced the trailing comma option for JavaScript
